cpp/autosar/src/rules/M5-14-1/RightHandOperandOfALogicalAndOperatorsContainSideEffects.ql

```diff
@@ -19,44 +19,12 @@ import codingstandards.cpp.autosar
 import codingstandards.cpp.SideEffect
 import codingstandards.cpp.sideeffect.DefaultEffects
 
-/**
- * an operator that does not evaluate its operand
- * `decltype` also has a non evaluated operand but cannot be used in a `BinaryLogicalOperation`
- */
-class UnevaluatedOperand extends Expr {
-  Expr operator;
-
-  UnevaluatedOperand() {
-    exists(SizeofExprOperator op | op.getExprOperand() = this |
-      not this.getUnderlyingType().(ArrayType).hasArraySize() and
-      operator = op
-    )
-    or
-    exists(NoExceptExpr e |
-      e.getExpr() = this and
-      operator = e
-    )
-    or
-    exists(TypeidOperator t |
-      t.getExpr() = this and
-      operator = t
-    )
-    or
-    exists(FunctionCall declval |
-      declval.getTarget().hasQualifiedName("std", "declval") and
-      declval.getAChild() = this and
-      operator = declval
-    )
-  }
-
-  Expr getOp() { result = operator }
-}
-
 from BinaryLogicalOperation op, Expr rhs
 where
   not isExcluded(op,
     SideEffects1Package::rightHandOperandOfALogicalAndOperatorsContainSideEffectsQuery()) and
   rhs = op.getRightOperand() and
   hasSideEffect(rhs) and
-  not exists(UnevaluatedOperand un | un.getOp() = rhs)
+  not rhs.(NoExceptExpr).getExpr().isUnevaluated() and
+  not rhs.(SizeofExprOperator).getExprOperand().isUnevaluated()
```

Comment:
There is no need to check whether the expressions of `NoExceptExpr` and `SizeofExprOperator` are unevaluated. That is always the case.

I think it is best to reinstate `UnevalutedOperand` and replace all but the `declval` case with `isUnevaluated` on the child of `rhs`.

We can then exclude the cases by asking if `not  rhs instanceof UnevaluatedOperand`

Reply:
agree! good suggestions